SPECS/tdnf/virtual-repo-snapshot.patch

```diff
@@ -0,0 +1,906 @@
+     char* pszCacheDir;
+     char* pszPersistDir;
+     char* pszProxy;
+diff --git a/solv/defines.h b/solv/defines.h
+index 38f5ab1..ebc7bd0 100644
+--- a/solv/defines.h
++++ b/solv/defines.h
+@@ -18,4 +18,32 @@
+         }                                                          \
+     } while(0)
+ 
++typedef struct {
++    // frequently changed values
++    char * pszElementBuffer;
++    int nBufferLen;
++    int nInPackage;
++    int nPrintPackage;
++    int nTimeFound;
++
++    // managed values
++    int nBufferMaxLen;
++    int nDepth;
++    int nPrevElement; // enum 0 -> start, 1 -> data, 2 -> end
++
++    //set and forget on creation
++    time_t nSearchTime;
++    FILE * pbOutfile;
++} TDNFFilterData;
++
++#define TDNF_MAX_FILTER_INPUT_THRESHOLD 10000000
++
++#define BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError) \
++    do {                                                           \
++        if (dwError)                                               \
++        {                                                          \
++            goto error;                                            \
++        }                                                          \
++    } while(0)
++
+ #endif /* __SOLV_DEFINES_H__ */
+diff --git a/solv/includes.h b/solv/includes.h
+index 2ab0c5c..1e39ecb 100644
+--- a/solv/includes.h
++++ b/solv/includes.h
+@@ -10,7 +10,6 @@
+ #include <unistd.h>
+ #include <stdbool.h>
+ #include <sys/types.h>
+-#include <sys/stat.h>
+ #include <dirent.h>
+ 
+ // libsolv
+@@ -44,4 +43,9 @@
+ #include "../history/history.h"
+ #include "prototypes.h"
+ 
++#include <expat.h>
++#include <stdlib.h>
++#include <time.h>
++#include <string.h>
++
+ #endif /* __SOLV_INCLUDES_H__ */
+diff --git a/solv/prototypes.h b/solv/prototypes.h
+index 2633b5e..0f5df8d 100644
+--- a/solv/prototypes.h
++++ b/solv/prototypes.h
+@@ -517,7 +517,8 @@ SolvReadYumRepo(
+     const char *pszPrimary,
+     const char *pszFilelists,
+     const char *pszUpdateinfo,
+-    const char *pszOther
++    const char *pszOther,
++    char * pszRepoEpochLimit
+     );
+ 
+ uint32_t
+diff --git a/solv/tdnfrepo.c b/solv/tdnfrepo.c
+index c27b907..aef6c61 100644
+--- a/solv/tdnfrepo.c
++++ b/solv/tdnfrepo.c
+@@ -9,6 +9,522 @@
+ 
+ #include "includes.h"
+ 
++// #### XML FILTER CODE ####
++void checkAndResizeBuffer(char ** pszCharBuffer, int * nBufferMaxLen, int nLengthToAdd) {
++    uint32_t dwError = 0;
++    int nBufferContentLen = strlen(*pszCharBuffer);
++    while (nBufferContentLen + nLengthToAdd >= *nBufferMaxLen)
++    {
++        *nBufferMaxLen *= 2;
++        char * pszTempCharBuffer = NULL;
++        pszTempCharBuffer = realloc(*pszCharBuffer, *nBufferMaxLen);
++        if (!pszTempCharBuffer)
++        {
++            dwError = ERROR_TDNF_TIME_FILTER_MEMORY;
++            BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++        }
++        //set expanded char buffer 
++        *pszCharBuffer = pszTempCharBuffer;
++    }
++cleanup:
++    return;
++error:
```

Comment:
Error handling for each function within the tdnf codebase is done using the BAIL_ON_ERROR/goto error structure. Each function expecting errors has an `error` and `cleanup` tag within. To go along with the existing method of handling errors I have added these tags to each of my functions and appropriately had the errors bail on each particular issue. 

That said, with this being brought into question and the return value above, I did convert most functions (sans the expat handler functions which are specified void function pointers) to return and bail on these dwError codes for consistency rather than having some be `void`, or NULL check a `char *`, etc. With there error printouts being placed in the top level sections on where the error can cascade up to using `pr_err` as done within other parts of the codebase.